### ISSUE TITLE
kata-ctl: Rewrite the kata-runtime utility program in rust

### ISSUE THREAD

**Issue by jodh-intel** [kata-containers/kata-containers#4499]
## TL;DR

Tracking project: https://github.com/orgs/kata-containers/projects/34/views/1

## Background

For Kata 2.x, we provide the [`kata-runtime` utility program](https://github.com/kata-containers/kata-containers/blob/main/docs/design/architecture/README.md#utility-program). Like the host side of Kata 2.x, this is written in golang.

## Upcoming changes

When the [`runtime-rs` PRs](https://github.com/kata-containers/kata-containers/pulls?q=is%3Apr+is%3Aopen+label%3Aruntime-rs) land in preparation for Kata 3.x, all components will be written in rust... except for the utility program.

## Rationale

We don't want to have to rely on "yet another toolchain" when we switch to a (mostly) pure rust environment:

- Extra complexity.
- Will increase the time to run the CI even further.

## Do we still need the utility program?

Yes, since it provides a number of useful features for users, admins and developers:

```bash
$ kata-runtime | sed -e '/./{H;$!d;}' -e 'x;/COMMANDS:/!d;'

COMMANDS:
   version            display version details
   check, kata-check  tests if system can run Kata Containers
   env, kata-env      display settings. Default to TOML
   exec               Enter into guest by debug console
   metrics            gather metrics associated with infrastructure used to run a sandbox
   factory            manage vm factory
   direct-volume      directly assign a volume to Kata Containers to manage
   iptables           get or set iptables within the Kata Containers guest
   help, h            Shows a list of commands or help for one command
```

> **Notes:**
>
> - `check` is semi-essential for users to allow them to determine if their hardware is capable of running Kata Containers.
> - `env` is very useful for debugging.
> - `exec` is an essential debug and development tool.

## Plan

- Rewrite [`kata-runtime`](https://github.com/kata-containers/kata-containers/tree/main/src/runtime/cmd/kata-runtime) in rust.
- Rename it `kata-ctl`.
  The current name is historical and arguably confusing for new users as it appears to be the runtime, but it isn't (see https://github.com/kata-containers/kata-containers/blob/main/docs/design/architecture/README.md#utility-program for further details on this).

## Thoughts

We _could_ consider merging `kata-runtime`/`kata-ctl` with the [`kata-agent-ctl`](https://github.com/kata-containers/kata-containers/tree/main/src/tools/agent-ctl) tool.

### Advantages

- Provide a single binary that is more functional that today's `kata-runtime`.
- No need to build the `agent-ctl` tool sepearately: the functionality would be provided in the standard package.

### Disadvantages

- Larger binary
  Not really a problem since this is a utility and **not** the actual shimv2 runtime binary ([`containerd-shim-kata-v2`](https://github.com/kata-containers/kata-containers/tree/main/docs/design/architecture#runtime)).

## Related

- https://github.com/kata-containers/kata-containers/issues/3917

**Comment by cmaf** [comment]
@jodh-intel I'm interested in this project; adding myself as an assignee

**Comment by jodh-intel** [comment]
Thanks @cmaf ! :smile:

**Comment by jodh-intel** [comment]
For those interested, we've started this work and are currently one of my branches:

- https://github.com/jodh-intel/kata-containers/tree/kata-ctl

Once we've got a minimal version of this tool working, we'll raise a PR to add it to the main repo.

**Comment by Ji-Xinyou** [comment]
Hi @jodh-intel , when `kata-ctl` is merged into main, could you give a roadmap here or in another issue,  because I would like to take part of the tasks if you guys need and it would be a mess if I just raise an issue separately.
